src/passes/DeadStoreElimination.cpp

```diff
@@ -0,0 +1,678 @@
+// a GC struct field N from a pointer P, after a store to that same pointer and
+// field. "Dynamic" here is used in the sense that we don't have a simple static
+// indexing of all the things we care about, sometimes called "lanes" in other
+// implementations. Instead we need to care about pointer identity, aliasing,
+// etc., which means we need to "dynamically" compare loads and stores and not
+// just a "lane" index computed for them. Note that in theory an index could
+// still be computed for such things, but probably at the cost of greater
+// complexity; if speed becomes an issue, then a refactoring in that direction
+// may be necessary. Such a refactoring might have downsides, however: While
+// having a single index for each operation is efficient, we would also need a
+// a way to represent "wildcards" - things that affect multiple indexes. For
+// example, an indirect call must be assumed to affect anything. But more
+// complex cases include a GC store of a certain type, which we can infer may
+// affect anything with a relevant subtype (but others cannot alias). To handle
+// that, we would need to store a set of indexes, already losing much of the
+// benefit of the indexed approach. Instead, the current code keeps things very
+// simple by just asking "may these two things interact", in which we can just
+// check the subtyping, etc.
+//
+
+#include <cfg/cfg-traversal.h>
+#include <ir/effects.h>
+#include <ir/local-graph.h>
+#include <ir/properties.h>
+#include <ir/replacer.h>
+#include <pass.h>
+#include <support/unique_deferring_queue.h>
+#include <wasm-builder.h>
+#include <wasm.h>
+
+namespace wasm {
+
+namespace {
+
+// A variation of LocalGraph that can also compare expressions to check for
+// their equivalence. Basic LocalGraph just looks at locals, while this class
+// goes further and looks at the structure of the expression, taking into
+// account fallthrough values and other factors, in order to handle common
+// cases of obviously-equivalent things. To achieve that, it needs to know the
+// pass options and features used, which we avoid adding to the basic
+// LocalGraph.
+struct ComparingLocalGraph : public LocalGraph {
+  PassOptions& passOptions;
+  FeatureSet features;
+
+  ComparingLocalGraph(Function* func,
+                      PassOptions& passOptions,
+                      FeatureSet features)
+    : LocalGraph(func), passOptions(passOptions), features(features) {}
+
+  // Check whether the values of two expressions are definitely identical. This
+  // is important for stores and loads that receive an input (like GC data),
+  // since we need to see that the pointer input is equivalent before we can
+  // tell if two stores overlap.
+  // TODO: move to LocalGraph if we find more users?
+  bool equivalent(Expression* a, Expression* b) {
+    a = Properties::getFallthrough(a, passOptions, features);
+    b = Properties::getFallthrough(b, passOptions, features);
```

Comment:
I might be mistaken, but for things such as trampling, is it sufficient to check the equivalency only in the fallthrough? What if a store is trampled in children? For example,
```wasm
(i32.load (i32.const 0))
(i32.store
  (block (result i32)
    ...
    (i32.store (i32.const 0) ...) ;; We don't check this?
    ...
    (i32.const 10)
  )
)
```